# Lesson 3: Pandas Part 1

This notebook is based on the official Pandas [documentation](https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html).  Pandas is a Python package that we will use to manage labeled data in R-like DataFrames.  Unless otherwise credited, quoted text comes from this document.  The official [10 Minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html#min) documentation is an excellent resource, and provides the titular guide which we will drawn from and expanded upon for this tutorial.




## Instructions
This tutorial provides step-by-step training divided into numbered sections. The sections often contain embeded exectable code for demonstration.  This tutorial is accompanied by a practice notebook: [L03-Pandas_Part1-Practice.ipynb](./L03-Pandas_Part1-Practice.ipynb). 

Throughout this tutorial sections labeled as "Tasks" are interspersed and indicated with the icon: ![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/16/Apps-gnome-info-icon.png). You should follow the instructions provided in these sections by performing them in the practice notebook.  When the tutorial is completed you can turn in the final practice notebook.  

## Introduction

*Where does this tool fit in my life?*

> For data scientists, working with data is typically divided into multiple stages: munging and cleaning data, analyzing / modeling it, then organizing the results of the analysis into a form suitable for plotting or tabular display. pandas is the ideal tool for all of these tasks.


*Why use this package?*

> pandas is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, real world data analysis in Python

### Pandas Cheat Sheet

Cheat sheets to remind you of popular Pandas functionality are available online.  You can find cheat sheets here:

+ [Data Quest Pandas cheat sheet](./media/pandas-cheet-sheet.pdf)
+ [Data Camp cheat sheet](./media/pandas-cheet-sheet.pdf)



---
## 1. Getting Started

First, we need to import the pandas library (and NumPy library too).  All packages are imported at the top of the notebook. Execute the code in the following cell to get started with this notebook.

In [1]:
# Numpy and pandas usage are often intertwined.
# These abbreviations are ubiquitiously used.
import numpy as np
import pandas as pd

The code above imports both NumPy and Pandas as variables named `np` and `pd`. We can use these variables to access the functionality of both libraries respectively.  Use of `np` and `pd` is commonly used by many. You will find online tutorials and documentation that use this convention. Therefore, we will use those object names for the rest of this class.

You may be wondering why we didn't import pandas like this:  
```python
import pandas
```
We could, but the first is far more commonly seen and the latter would require that we prefix all calls to pandas fuction with `pandas.`. It is easier to just prefix with `pd.`. 

#### Task 1a: Setup

<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

In the practice notebook, import the following packages:
+ `numpy` as `np`
+ `pandas` as `pd`


---
## 2. Data Object Creation

Here we will learn to create and use the two data structures provided by Pandas: **Series** and **DataFrames**.  For additional help, see the [Data Structure Intro section](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dsintro).


In short, Pandas provides two new objects in which to store and manipulate data.

Dimensions | Name |	Description
-----------|-------|--------------------------------------
1 |	Series | 	1D labeled homogeneously-typed array
2 |	DataFrame | 	General 2D labeled, size-mutable tabular structure with potentially heterogeneously-typed column

Typical usage of this Pandas with data analysis will involve the creation of one or more `DataFrame` objects, and manipulating them with the provided methods.

### 2.1 The `pd.Series` Data Object

In Pandas, a Series is a one dimensional array, and similar to NumPy, all the values are of the same data type (i.e. integer, string, boolean, etc.) or **NaN**.  NaN means "not a number" but serves as a marker when a value is missing.  

The following cell creates a new series containing an array of 5 numbers and one missing value by simply passing a list of 6 elements to the `pd.Series` function. Execute the cell:

In [2]:
# Create the series object.
my_series = pd.Series([1, 3, 5, np.nan, 6, 8])

Notice that we have introduced a missing values by using the NumPy variable `np.nan`. 

Execute the next cell to see what the new series object named `my_series` contains

In [3]:
# Call the object in the notebook.
# This (mostly) calls the __repr__() function of a given object.
my_series

0    1.0
1    3.0
2    5.0
3    NaN
4    6.0
5    8.0
dtype: float64

#### Task 2a Create a `pd.Series` object
<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>
Use the practice notebook perform the following:
+ Create a series of your own design.

### 2.2 The `pd.DataFrame` Data Object

In Pandas, a DataFrame is a two-dimensional array.  Just like NumPy, it has axes labeled "rows" and "cols".  However, unlike NumPy, Pandas DataFrames have a few advantages over Numpy 2D matrices:  
1. DataFrames allows for different data types in each column, where as NumPy arrays must always be of the same data type.
2. DataFrames provide greater flexibilty for indexing and slicing.

#### 2.2.1 Create a DataFrame with a dictionary

A dataframe is created by passing a Python 2D dictionary (i.e. **dict**) to the `pd.DataFrame` function.  Remember dictionaries have key/value pairs. The dictionary is interpreted in the following way:

- The keys in the dictionary argument becomes the column names
- The value in the dictionary argument contains the column values
- Each "column" must have the same number of values.

The following cell contains code that creates a new DataFrame object named `df_1`.  This dataframe consists of two columns, one named "alpha" and the other "beta".  Each column has 5 elements. Notice that the columns contain different data types.  Execute the following cell.

In [4]:
# Creating a dataframe with a dictionary.
df_1 = pd.DataFrame(
    {'alpha': [0, 1, 2, 3, 4],
     'beta': ['a', 'b', 'c', 'd', 'e']})

Execute the next cell to see what the new data frame object named `df_1` contains

In [5]:
# Call the object in the notebook.
# This time we get a html output, as the notebook shell
# sees a _repr_html_() function to call.
df_1

alpha beta
0      0    a
1      1    b
2      2    c
3      3    d
4      4    e

Observe that a DataFrame can be visualized as a data table. Importantly, note that, a numerical index has been automatically generated and is shown in the left-most column.

#### 2.2.2. Create a DataFrame with a Numpy array
Alternatively, we can create a DataFrame by passing in a Numpy matrix.  Remember that Numpy arrays can only store one data type, so if the data contains mixed types you would not use this approach. 

The following code creates a Dataframe of 6 rows and 4 columns and fills it with random numbers generated using the Numpy `np.random.randn` function.  Execute the following cell:

In [6]:
df = pd.DataFrame(np.random.randn(6, 4))

Let's take a look at our new data frame by executing the following cell:

In [7]:
df

0         1         2         3
0  0.575863  0.955110 -0.720598 -0.061962
1 -0.187312  0.955657 -0.914807  0.239739
2 -0.247863 -0.963321 -0.033706  0.194691
3  0.223908  0.648303 -1.545000  1.474062
4 -0.333893 -0.033576  0.158809 -0.508554
5 -0.363903 -0.754240 -1.104878  0.068556

Observe that because we did not provide a dictionary, we do not have header names for each column. Rather a numeric column index was automatically generated.

#### Task 2b: Creating a DataFrame

<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

In the practice notebook perform the following:
+ Create a pd.DataFrame object from a Python dictionary. Design the data as you like.

### 2.3. Create a DataFrame with row and column labels

Rather than have a numeric index for each row and column of the DataFrame, we may want to give each row a human-readable label. This will help make the data more accessible (as will be seen later).  To demonstrate, lets reuse the 6x4 datafame we created in section 2.2.2. Using this dataset we want to rename the row indexes with hypothetical dates that each row of information was collected, and we want to give the columns names, such as `A1`, `A2`, `B1` and `C1`.  

First, let's set the row index names.  The code below uses the `pd.date_range` function to create a list of dates. Most likely, if we have dates in a real input dataset we would not be creating a list of dates, but for the purposes of demonstration we will generate a series of regularly spaces dates over a 6 day period: one for each of the 6 rows. Note that the `start` argument for the `pd.date_range` function expects the date in YYYYMMDD format.

In [8]:
dates = pd.date_range('20200301', periods=6)
dates

DatetimeIndex(['2020-03-01', '2020-03-02', '2020-03-03', '2020-03-04',
               '2020-03-05', '2020-03-06'],
              dtype='datetime64[ns]', freq='D')

Now, we can re-create the DataFrame from section 2.2.2 with dates as row indexes and  `A1`, `A2`, `B1` and `C1` as column names. We do this by using the `index` and `columns` arguments when creating the DataFrame as shown in the following cell.  Execute the cell to see the results: 

In [9]:
df = pd.DataFrame(np.random.randn(6, 4), index=dates, columns=('A1','A2', 'B1', 'C1'))
df

A1        A2        B1        C1
2020-03-01 -0.497505 -0.389780  0.431971  0.080003
2020-03-02  0.159576 -0.424171  1.267680 -2.124108
2020-03-03  0.523217 -0.537805  0.647918 -0.337616
2020-03-04  0.909497  0.525794 -0.262309  0.060207
2020-03-05  0.293617  0.359641  0.455156 -0.366412
2020-03-06  0.489890  0.035871  1.168338 -0.818020

#### Task 2c: Create DataFrame with labels

<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

In the practice notebook perform the following:

+ Create a 10x5 dataframe of random numeric integers that follow a [Guassian (normal) Distribution](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.normal.html). 
  + Center the distrubtion at 0.85.
  + We will use these values as assumed grades for a class of students
+ Adjust the row indexes to be the names of hypothetical students.
+ Adjust the columsn to be the names of hypothetical projects, homework, exam names, etc.


---
## 3. Loading Data

Creating a DataFrame manually, as in the previous examples, would be tedious for large datasets.  Fortunately, Pandas provides a variety of functions for importing data from files.  This saves you from writing your own Python code to open a file, read in the contents and format it for inclusiong into a DataFrame. For this tutorial we will look at two of these import functions:  `pd.read_csv` and `pd.read_excel`.

For this portion of the tutorial, we will use a rather [famous set of data concerning iris flowers](https://en.wikipedia.org/wiki/Iris_flower_data_set).  You can find a copy of the comma-separated file accompanying this notebook in the `data` directory. You can view the file [here](./data/iris.csv).

![Iris flower](https://upload.wikimedia.org/wikipedia/commons/thumb/4/41/Iris_versicolor_3.jpg/193px-Iris_versicolor_3.jpg)


### 3.1 Importing Using `pd.read_csv`

The `pd.read_csv` function can be used to read in files that are in comma-separated format or tab delimited format: two very common data formats.  This function is very flexible and has a large number of arguments. See the [`pd.read_csv` online documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) for a thorough listing of arguments. 

The iris data is in CSV format and is compatible with the default arguments for the `pd.read_csv` function, therefore we can easily import this data by providing the path to the file as the only argument to the function. This will automatically create a DataFrame for us! Execute the following line to import the data:

In [10]:
iris_df = pd.read_csv('data/iris.csv')

#### Task 3a: Import the iris.csv file
<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

In the practice notebook:
+ Import the iris dataset.
+ Take a look at the `pd.read_csv` online documentation. Write example code in a Markup cell for how you would import this file if it were tab-delimited.




### 3.2 Importing using `pd.read_excel()`
It is common for data to be stored in Excel worksheets (provided the data is not too large).  You can easily export data from Excel to CSV or Tab delimited formats but Pandas provides  the `pd.read_excel` function to save you the time.  We will not practice loading data from Excel, but take a look at the [`read_excel` online documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html) for details.

---
## 4. Explore your Data
There are a variety of functions for exploring data once you have imported it. Here we will discuss `pd.head`, `pd.tail`, `pd.sample`, and `pd.describe`.

### 4.1 The `pd.head` function

As with any Python variable, you can display its contents by simply typing the variable name and pressing the "enter" key.  Howevever, for very large data files this may not be desired.  The data used in this tutorial is not extremly large but for large data we need better ways to examine the data.

To easily examine the first set of rows in the DataFrame, use the `head` function of the `pd.DataFrame` object. By deafult it displays the first 5 rows. Execute the following cells to view the first five rows of the iris dataset.

In [11]:
iris_df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

This is an excellent way to check if you created your `pd.DataFrame` object correctly.  You can change the number of rows returned by providing an argument.  Note, in the table displayed above that the row index have been generated as numeric values but the column indexes are the names of the column headers from the input data file.  If the file has headers, the `pd.read_csv` function will automatically use those as column index names.


The following requests the first 10 rows of the data frame:

In [12]:
iris_df.head(10)

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa
5           5.4          3.9           1.7          0.4  setosa
6           4.6          3.4           1.4          0.3  setosa
7           5.0          3.4           1.5          0.2  setosa
8           4.4          2.9           1.4          0.2  setosa
9           4.9          3.1           1.5          0.1  setosa

### 4.2 The `pd.tail` function
Simliar to the `pd.head` function, the `pd.tail` function lets you peak at the last 5 rows of data.  To demonstrate, execute the following cell.

In [13]:
iris_df.tail()

sepal_length  sepal_width  petal_length  petal_width    species
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

### 4.3 The `pd.sample` function
The `pd.sample` function selects a random set of rows to display. By default it only selects 1 row, we can provide the number of rows we would like (e.g. 5) by providing the number as an arugment:

In [14]:
iris_df.sample(5)

sepal_length  sepal_width  petal_length  petal_width     species
100           6.3          3.3           6.0          2.5   virginica
7             5.0          3.4           1.5          0.2      setosa
97            6.2          2.9           4.3          1.3  versicolor
129           7.2          3.0           5.8          1.6   virginica
76            6.8          2.8           4.8          1.4  versicolor

Repeat execution of the function to see a different set of randomly selected rows:

In [ ]:
iris_df.sample(5)

Note, the `sample` function can be very important beyond just viewing random rows. If you ever need to select a random sample of records from a dataset the `sample` function can do that for you.

### 4.4 The `pd.describe` function
The `pd.describe` function helps demonstrate the power of Pandas over Python dictionaries and Numpy arrays.  It provides a summary of statistics for the data frame in one simple function call. These statistics include the count, mean, standard deviation, min, max and quartiles.

Execute the following cell to demonstrate:

In [15]:
iris_df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

#### Task 4a: Explore Data

<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

In the practice notebook:

 + Use `head`, `tail` and `sample` with the iris dataset.
 + Do the same with the dataset you created in task 2c.


---
## 5. Accessing Data

In this section we will learn to retreive values, **slice** the data frame, and use column and row labels (rather than numeric indexes) to find data. 

### 5.1 Viewing column and row labels.
Before we delve into retreiving data, lets first look at some properties of DataFrames for retrieving the column and row index names.  It is convenient to work with, or check the indexes and columns of a data frame using their names rather than a numeric index.  First, to retrieve the columns names use the `columns` property of the DataFrame.

In [16]:
iris_df.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')

Similary, we can get a description of the index using the `index` property:

In [17]:
iris_df.index

RangeIndex(start=0, stop=150, step=1)

Observe that here our index is numeric and runs from 0 to 150, incremented by 1

#### Task 5a: Viewing columns and rows

<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

In the practice notebook:
+ Display the columns and indexes of the iris dataset.
+ Do the same with the dataset you created in Task 2c.


### 5.2. Get all the values

Perhaps you would like to extract the contents of the DataFrame into a 2-dimensional Python list. You will lose the power of Pandas, but you should know how to do this. There are two ways to extract values: the new Pandas > v0.24 way and the older way.  

This following code example used to be the standard way of getting the values of a dataframe as a NumPy array:

```python
my_array = df.values
```

This is depreciated in version 0.24 of Pandas to instead require the use of `to_dict` function.

```python
my_array = df.to_dict()
```

You can verify what version of `pandas` you have:

In [18]:
pd.__version__

'0.25.1'

#### Task 5b: Get Values

<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

In the practice notebook: 

+ Check the version of `pandas` you have 
+ Use the appropriate method to convert the iris data to a dictionary.
+ Do the same with the dataset you created in Task 2c.


### 5.3 Basic Selections
Pandas provides a variety of row/column selectors that are similar to the way that lists and arrays are indxed in Python and NumPy. This tutorial will introduce those. 


#### 5.3.1 Get a column
The easiest way to select a single column is to use the column label! When in doubt, use the `.column` property of the DataFrame to find the list of column labels.  Using the column label will result in the column being returned as a Series object.  For example, if we want the `sepal_length` column of the iris data frame we can execute the following:

In [ ]:
slength = iris_df['sepal_length']
slength.head()

Observe, we can use the `.head` function to get the first few rows of a Series as well as a DataFrame.

#### 5.3.2 Get rows (slicing)
Similar to slicing of Python lists, we can do the same with Pandas DataFrames:

```python
iris_df[start:end]
```

As a reminder, indexing of a list in Python uses 0-based indexing and negative numbers index in reverse order:

```
 +---+---+---+---+---+---+
 | P | y | t | h | o | n |
 +---+---+---+---+---+---+
   0   1   2   3   4   5  
  -6  -5  -4  -3  -2  -1
```

**Important:** Recall the recommendation to *not* using this type of indexing (slicing) in production with very large data. Instead use other provided functions (`.loc`, `iloc`, etc). However, to practice, let's get the first row of the iris dataset by slicing:

In [19]:
iris_df[0:1]

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa

Next lets get the first five rows:

In [20]:
iris_df[0:5] 

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

### 5.4. Selecting with `df.loc` and `df.iloc`

Aside from the column name and slicing (as shown in the previous sections) Pandas provides the `df.loc` and `df.iloc` accessor.  These accessors perform better for selecting subsets of the dataframe. The `.loc` accessor is used to select rows and columns using index names and `.iloc` is used for selecting by numeric indexes.  Although, if your DataFrame only has numeric row indexes then `.loc` can still be used.

However, please note:
> While standard Python / NumPy expressions for selecting and setting are intuitive and come in handy for interactive work, for production code, we recommend the optimized pandas data access methods, '.at', '.iat', '.loc' and '.iloc'.

In other words, rather than index Pandas arrays using something like, `iris[1:5,]`, instead use the functions that will be described below.  They are much faster and optimzied for larger datasets.

### 5.4.1 Using `.loc`

The `.loc` accessor allows for the slicing of the DataFrame by using the labels of either rows and columns.  

The syntax is as follows:

```python
df.loc[row_start:row_end, column_start:column_end]
```

Remember, because the rows in the iris data frame are indexed using integers, we can use `.loc` to get rows using the integer values.  To get the first row:

In [21]:
iris_df.loc[0]

sepal_length       5.1
sepal_width        3.5
petal_length       1.4
petal_width        0.2
species         setosa
Name: 0, dtype: object

To get the first two rows:

In [22]:
iris_df.loc[0:1]  # Get the first two rows.

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa

Observe that when only one row is returned that a Series is provided.  When multiple rows are returned then a new DataFrame object is returned.


Next, recall that we constructed a 6x4 dataframe in section 2.3 that was filled with random values and we set the row names using datetime values. The DataFrame was named `df`.  As a reminder here are the first 2 rows:

In [23]:
df.head(2)

A1        A2        B1        C1
2020-03-01 -0.497505 -0.389780  0.431971  0.080003
2020-03-02  0.159576 -0.424171  1.267680 -2.124108

In this data frame the rows are indexed using strings that represent dates. We can therefore use those labels to slice the dataframe.  The following extracts the first row:

In [24]:
df.loc['2020-03-01']

A1   -0.497505
A2   -0.389780
B1    0.431971
C1    0.080003
Name: 2020-03-01 00:00:00, dtype: float64

Similarly we can get the first two rows:

In [25]:
df.loc['2020-03-01':'2020-03-02']

A1        A2        B1        C1
2020-03-01 -0.497505 -0.389780  0.431971  0.080003
2020-03-02  0.159576 -0.424171  1.267680 -2.124108

#### 5.4.2 Using `.loc` to get columns
Remember for our iris data frame the column indexes have the names as provided by the `.columns` property.

```python
iris_df.columns
Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')
```

We can use these names to retrive values for specific columns. For example, the code below extracts rows 4 through 6 and the columns 'sepal_width' and 'petal_width':

In [26]:
iris_df.loc[4:6, 'sepal_width': 'petal_width']

sepal_width  petal_length  petal_width
4          3.6           1.4          0.2
5          3.9           1.7          0.4
6          3.4           1.4          0.3

The use of the colon in the code above between 'sepal_width' and 'petal_width' implies a range. Therefore every column between those two are included. This is why 'petal_length' appears in the output.  If we wanted to specifically select only the two columns and none other, we could provide the names in a list:

In [27]:
iris_df.loc[4:6, ['sepal_width', 'petal_width']]

sepal_width  petal_width
4          3.6          0.2
5          3.9          0.4
6          3.4          0.3

Observe that now we only have 'sepal_width', and 'petal_width' in the resulting data frame.

Moreover, we can limit the rows to only those specified in the same way:

In [28]:
iris_df.loc[[4,6], ['sepal_width', 'petal_width']]

sepal_width  petal_width
4          3.6          0.2
6          3.4          0.3

#### Task 5c: Using `loc`

<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

In the practice notebooks:

+ Use any iris dataframe to:
  + Select a row slice with `loc`.
  + Select a row and column slice with `loc`.
  + Take a look at the [Pandas documentation for the `at` selector](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.at.html). Use what you learn there to select a single item with Pandas `at` accessor.

### 5.4.2 Using `.iloc`

The `.iloc` accessor allows us to slice the data frame using an integer index, regardless of what the index (or column) labels actually are.

The syntax is the same as `.loc` with the exception of that integers are used instead of labels:

```python
df.iloc[row_start:row_end, column_start:column_end]
```
To retrieve the first row of the iris data we use the numeric index:

In [29]:
iris_df.iloc[0]  # Get the first row.

sepal_length       5.1
sepal_width        3.5
petal_length       1.4
petal_width        0.2
species         setosa
Name: 0, dtype: object

We can use any combination of integer indexes to get a subset of rows and columns.  The following examples retrieves rows 2 to 5, and excludes the 'species' column.

In [32]:
iris_df.iloc[2:5, :-1] 

sepal_length  sepal_width  petal_length  petal_width
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

#### Task 5d: Using `iloc`

<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

In the practice notebooks:

+ Use any iris dataframe to:    
    + Select a row slice with `iloc`.
    + Select a row and column slice with `iloc`.
    + Take a look at the [Pandas documentation for the `iat` selector](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iat.html). Use what you learn there to select a single item with Pandas `iat` accessor.


### 5.5 Boolean Indexing

Pandas allows us to use boolean values to extract specific rows of data that meet certain conditions.To clarify this, lets first examine what occurs when we apply a condition to a data frame.  Create a new dataframe with the following

In [33]:
df_2 = pd.DataFrame(
    {'alpha': [0, 1, 2, 3, 4],
     'beta': [5, 6, 7, 8, 9]})
df_2

alpha  beta
0      0     5
1      1     6
2      2     7
3      3     8
4      4     9

Let's apply a condition to the data frame to find all values > 1.0

In [34]:
df_2 > 1.0

alpha  beta
0  False  True
1  False  True
2   True  True
3   True  True
4   True  True

Notice that the resulting data frame contains `True` and `False` values in place of every value that did or did not meet the condition.  We can use this new data frame to subset the original dataframe.

Lets consider the iris dataset for a more realistic example.  Suppose we want to find all rows with a `sepal_length` greater than 5.8. We know from the call to `describe` that the mean is approximately 5.8. First, we will identify rows that meet our criteria:

In [35]:
condition = iris_df['sepal_length'] > 5.8
condition

0      False
1      False
2      False
3      False
4      False
       ...  
145     True
146     True
147     True
148     True
149     True
Name: sepal_length, Length: 150, dtype: bool

Notice that the Series `condition` indicates `False` if the sepal_length value in the row is <= 5.8 and `True` if it is > 5.8. We can use this array to slice the rows of the dataframe in the following way (only rows with a `True` in the `condition` object are returned):

In [36]:
iris_df[condition].head()

sepal_length  sepal_width  petal_length  petal_width     species
50           7.0          3.2           4.7          1.4  versicolor
51           6.4          3.2           4.5          1.5  versicolor
52           6.9          3.1           4.9          1.5  versicolor
54           6.5          2.8           4.6          1.5  versicolor
56           6.3          3.3           4.7          1.6  versicolor

You could rewrite the lines above to combine them into a single line with identical results:

In [37]:
iris_df[iris_df['sepal_length'] > 5.8].head()

sepal_length  sepal_width  petal_length  petal_width     species
50           7.0          3.2           4.7          1.4  versicolor
51           6.4          3.2           4.5          1.5  versicolor
52           6.9          3.1           4.9          1.5  versicolor
54           6.5          2.8           4.6          1.5  versicolor
56           6.3          3.3           4.7          1.6  versicolor

#### Task 5e: Boolean Indexing

<span style="float:right; margin-left:10px; clear:both;">![Task](http://icons.iconarchive.com/icons/sbstnblnd/plateau/96/Apps-gnome-info-icon.png)
</span>

In the practice notebook:

+ Create subsets of the iris dataset using boolean indexes that:
    + Use one boolean operator.
    + Use two boolean operators.
